File: src/api/ffmpeg.ts
import { invoke } from '@tauri-apps/api/core';
import { listen } from '@tauri-apps/api/event';
import type { FFmpegStatus, DownloadProgress } from '../types/ffmpeg';

export async function checkFFmpegStatus(): Promise<FFmpegStatus> {
    return await invoke<FFmpegStatus>('check_ffmpeg_status');
}

export async function downloadAndInstallFFmpeg(onProgress: (progress: number) => void): Promise<void> {
    const unlisten = await listen<DownloadProgress>('ffmpeg-download-progress', (event) => {
        onProgress(event.payload.percentage);
    });

    try {
        await invoke('download_and_install_ffmpeg_command');
    } finally {
        unlisten();
    }
}

File: src-tauri/src/commands/ffmpeg.rs
use crate::services::{download_and_install_ffmpeg, get_ffmpeg_status};
use crate::types::FFmpegStatus;
use tauri::Window;

#[tauri::command]
pub async fn check_ffmpeg_status() -> Result<FFmpegStatus, String> {
    Ok(get_ffmpeg_status())
}

#[tauri::command]
pub async fn download_and_install_ffmpeg_command(
    window: Window,
) -> Result<(), String> {
    download_and_install_ffmpeg(window).await
}
